internal/http/parser/http_response.go

```diff
@@ -0,0 +1,225 @@
+package http_parser
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+)
+
+type Header struct {
+	Key   string
+	Value string
+}
+
+type StatusLine struct {
+	Version    string
+	StatusCode int
+	Reason     string
+}
+
+type HTTPResponse struct {
+	StatusLine StatusLine
+
+	Headers []Header
+
+	Body []byte
+}
+
+func Parse(data []byte) (httpResponse HTTPResponse, readBytesCount int, err error) {
+	reader := bytes.NewReader(data)
+
+	response, err := doParseResponse(reader)
+	if err != nil {
+		return HTTPResponse{}, 0, err
+	}
+
+	return response, len(data) - reader.Len(), nil
+}
+
+func doParseResponse(reader *bytes.Reader) (HTTPResponse, error) {
+	var allHeadersFound = false
+	var sectionsFound int
+	var R HTTPResponse
+	var SL StatusLine
+
+	offsetBeforeCurrentSection := GetReaderOffset(reader)
+	versionB, err := ReadUntilAnyDelimiter(reader, [][]byte{SPACE, TAB})
+	if err == io.EOF {
+		return HTTPResponse{}, IncompleteInputError{
+			Reader:  reader,
+			Message: fmt.Sprintf("Expected: HTTP-version, Received: %q", versionB),
+		}
+	}
+	version := string(versionB)
+	// HTTP/1.X
+	if len(version) != 8 {
+		// Seek to ending point of last sub-section
+		reader.Seek(int64(-2), io.SeekCurrent)
+		return HTTPResponse{}, BadProtocolError{
+			Reader:  reader,
+			Message: "Invalid HTTP-version field length",
```

Comment:
Bad error, users won't know what a "HTTP-version field" is - just say "Expected HTTP/1.1"